fix(helpers): kernel_symbols change and fix

rafaeldtinoco opened the pull request:
The whole "lazy" concept for the kallsyms file relies in the concept that symbols would have a single address only (and the assumption that the file is mostly sorted by symbol addresses).

Checking kernel_symbols.go and the KernelSymbolTable interface, one could change GetSymbolByName to return a slice of []*KernelSymbols and that would be an easy change for fullKernelSystemTable.

Problem is that the lazy implementation relies in stopping to read the kallsym file once a symbol is picked, or in a binary search of address considering file is sorted, etc. That doesn't work well for the same symbol having more than 1 address.

Quickly checking kallsyms for duplicate symbol addresses, the generic names will have huge amount of duplicates:

1693 __func__.0
1198 _entry.1
834 __func__.2
777 _entry.3
...

and, the "unique kernel symbols" will have 2 or 3 addresses (under certain circumstances, like when the symbol is static to a source file, or under certain compilation optimizations):

2 switch_mm
2 sw_fence_dummy_notify
2 suspend_attrs
2 suspend_attr_group
2 subsystem_id_show
2 str__i915__trace_system_name
...

There is also the case where the same address has multiple symbols:

ffffffffc0310200 b __key.22	[drm_display_helper]
...
ffffffffc0310200 b __key.17	[drm_display_helper]
ffffffffc0310200 b drm_dp_aux_dev_class	[drm_display_helper] ...

So in both cases, when indexing by sym name, or by sym address, code should account for the possibility of having multiple results.

This change makes the helper "fast enough" while allowing it to return multiple values from its maps.

Related: https://github.com/aquasecurity/tracee/issues/3798

geyslan reviewed (approved):
@rafaeldtinoco this is an amazing work. LGTM.

I've put some comments.

rafaeldtinoco commented on helpers/kernel_symbols.go:
```diff
@@ -3,340 +3,315 @@ package helpers
-		return symbol, nil
+	symbolOwner := "system"
+	if len(line) > 3 {
+		symbolOwner = strings.Clone(line[3])
```
Alright. Will push the changes soon.

rafaeldtinoco commented on helpers/kernel_symbols.go:
```diff
@@ -3,340 +3,315 @@ package helpers
-func symbolKey(owner, name string) string {
-	return owner + "_" + name
+type addrAndOwner struct {
+	addr  uint64
```
I dont want a struct inside another struct as a map key, thats why.

rafaeldtinoco commented on helpers/kernel_symbols.go:
```diff
@@ -3,340 +3,315 @@ package helpers
+	k.updateLock.RLock()
+	defer k.updateLock.RUnlock()
+
+	symbols, exist := k.symByName[nameAndOwner{n, o}]
```
well this styling preference, the variable names tell what they are and its way more concise not to use them, and they're private.

rafaeldtinoco commented on helpers/kernel_symbols.go:
```diff
@@ -3,340 +3,315 @@ package helpers
+// mono-threaded implementation that parses + processes the lines takes ~0.700
+// seconds. This approach takes ~0.350 seconds (2x speedup).
+//
+// NOTE: The procfs file reading cannot be paralelized because procfs does not
+// implement mmap (so the reading could be split into chunks and processed in
+// parallel).
```
I rephrased the last part.

rafaeldtinoco commented on helpers/kernel_symbols.go:
```diff
@@ -3,340 +3,315 @@ package helpers
-		line := strings.Fields(line)
-		// if the line is less than 3 words, we can't parse it (one or more fields missing)
-		// if the searched owner isn't system and the line counter is less than 4 words, the line is irrelevant
-		if len(line) < 3 || (owner != "system" && len(line) < 4) {
```
Not sure, but I dont think all kallsyms variations were considered when this package was created and this was a generic attempt of not failing. I think the original idea was possibly to avoid the final lines, coming from modules (thus not "system"), to break something, but can't tell for sure and haven't seen a real scenario it was needed.

rafaeldtinoco commented on helpers/kernel_symbols.go:
```diff
@@ -3,340 +3,315 @@ package helpers
+// updates all maps OR multiple goroutines with fine grained locking (turned out
+// to be slower than a single goroutine).
+//
+// Simple file parsing (no processing) takes ~0.200 seconds on a 4-core machine.
+// If buffer is increased to 4MB, it might take ~0.150 seconds. A simple
+// mono-threaded implementation that parses + processes the lines takes ~0.700
+// seconds. This approach takes ~0.350 seconds (2x speedup).
```
I'm not doing performance tuning/benchmarking for this change (since Im actually removing one attempt to do so that caused the original problem because of wrong premises). Be my guest to open an issue and benchmark any optimization to this code =).

rafaeldtinoco commented:
Thanks for the careful review @geyslan. I think I have addressed most of your comments! I'll merge once the new push passes the tests and then focus in the tracee changes.
